### ISSUE TITLE
Add Async pattern for TokenValidation

### ISSUE THREAD

**Issue by MicahZoltu** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#468]
There are a number of places where `JWTSecurityTokenHandler` calls into user provided methods that may be doing some IO (e.g., database or remote calls) but they do so in a synchronous way.

One example of such a place is with JWT signature validation.  The user can provide their own `SignatureValidator` or `IssuerSigningKeyResolver`.  The act of validating a signature or resolving signing keys may require an interaction with an external server.

For a real-world example, Google rotates its signing keys regularly (rumor is daily) and you can get the latest public key in JWK format [here](https://www.googleapis.com/oauth2/v3/certs).  Unfortunately, it is not possible to know when they are going to roll their keys so an application that is attempting to validate Google issued JWTs will need to hit that endpoint to retrieve the latest  signing key.  The application could be optimized to cache the signing keys but that is still an async operation and it is entirely up to Google as to how often they rotate their keys (they could start rotating them for every request if they wanted).

Another real-world example once again with Google's authentication, Google supports sending the token to [here](https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=XYZ123) for validation rather rather than doing the validation locally.  If one wants to use this mechanism then every token validation would be a remote call.

Unfortunately, I recognize that the current API is very async un-friendly and the interfaces it implements are also not async-friendly but I wanted to get this filed anyway in hopes that at some point things can be improved to support modern asynchronous use cases.

**Comment by brentschmaltz** [comment]
@Zoltu I agree with you that off-box validation is a real scenario. We should have an async pattern up and down the call graph to realize this pattern.
We stopped at metadata retrieval.

**Comment by brentschmaltz** [comment]
We will need to add an new return type: ValidationResult so the 'out' param is not needed.

**Comment by brentschmaltz** [comment]
Include this issue with this fix.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/566

**Comment by brentschmaltz** [comment]
Modified title to reflect that this is general in nature across all token types. Perhaps in SecurityTokenHandler

**Comment by shagenhokie** [comment]
Do we have an ETA on this?

**Comment by brentschmaltz** [comment]
@shagenhokie not yet

**Comment by fragilethoughts** [comment]
@brentschmaltz the method that handles authentication is async so I think it shouldn't be that difficult to add? https://github.com/aspnet/Security/blob/beaa2b443d46ef8adaf5c2a89eb475e1893037c2/src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs#L107-L128

**Comment by brentschmaltz** [comment]
@fragilethoughts I agree we need async, particularly with delegation to validation services, such as KeyVault and Google. Not only do we need ValidateTokenAsync, we need async on our Crypto such as SignatureProviders and KeyWrapProviders.

This is a feature we are thinking about for 5.3.1, but realistically it may be later. The other issue is we want to remove references to any libraries other than those shipped by Microsoft, such as JSON.net. With many JwtSecurityToken and JwtSecurityTokenHandler API's using JSON.net, we created a new assembly M.IM.JsonWebTokens where ValidateToken returns a structure TokenValidationResult and can have easily have an async version. We are free to invent here.

**Comment by fragilethoughts** [comment]
Sounds good

**Comment by dhjf** [comment]
Is the plan to create an async version of JwtSecurityTokenHandler, so it would be possible to call WriteToken with asynchronously? It would help a lot, being able to go all async using IdentityServer4 for signing access tokens.

**Comment by mcthuesen** [comment]
Any status on this? 

We have this issue as well where we need to do an http request to an HSM to sign access tokens. Our solution right now is to use [WebClient.cs](https://docs.microsoft.com/en-us/dotnet/api/system.net.webclient?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev15.query%3FappId%3DDev15IDEF1%26l%3DEN-US%26k%3Dk(System.Net.WebClient);k(DevLang-csharp)%26rd%3Dtrue&view=netframework-4.7.2) instead of the recommended way of doing http requests using [HttpClient.cs](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netframework-4.7.2). HttpClient does not support synchronous requests.

**Comment by brentschmaltz** [comment]
@mcthuesen 	We haven't moved forward much on this, but it is important for a number of reasons. We are not sure if this fits into our 5.x release or will need to wait for 6.0.
Our 6.0 release fits with CORE 3.0.

**Comment by mfeingol** [comment]
Speaking on behalf of Azure DevOps, we'd love to have an async variant of the `AudienceValidator` delegate.

**Comment by LeroyK** [comment]
Any idea when this will be available?

**Comment by brentschmaltz** [comment]
@LeroyK at the earliest in the 1st qtr of 2020 for a full stack.
However if Async Delegates were available would that satisfy a pent up need?

**Comment by daniel-munch-cko** [comment]
Just came across this, was there any follow-up? I'm mostly interested in async versions of `SignatureProvider.Sign` and `SignatureProvider.Verify`

**Comment by brentschmaltz** [comment]
@daniel-munch-cko we are working on shipping 6.x where we have a number of internal fixes being validated. Once that is out, we will address this as a lot of interest has been expressed. Particularly where crypto operations are performed out-of-process.

**Comment by stebet** [comment]
This would be extremely helpful for our case where we have an ecosystem of services that communicate with each others APIs using JWT tokens, so they need to fetch signing keys on demand as they are rotated.

**Comment by cyberhck** [comment]
@stebet that's exactly my usecase as well, it sucks having to do `.Result` on `async` method

**Comment by stebet** [comment]
> @stebet that's exactly my usecase as well, it sucks having to do `.Result` on `async` method

Shhhh... you'll invoke @davidfowl....

**Comment by brentschmaltz** [comment]
@cyberhck @stebet @daniel-munch-cko yes we hear you. We have some internal work to do with our 6.x branch (our dev is ahead of dev5x). We do not want to backport this work, so the order of items is 6.x public, coordinating with asp.net, identityserver to ensure no breaking issues. 
Then we can address these long standing issues.

**Comment by bugproof** [comment]
It's open since 2018. Do you need help? I can do it in 10 minutes but with a breaking change in the API, or do you want it to be backwards compatible?

**Comment by bugproof** [comment]
Kinda funny to see it open since 2016.

**Comment by brentschmaltz** [comment]
@bugproof fair enough. When opened the win was seen as a programming convenience, in a large number of cases, as almost all crypto operations were inproc and synchronous. That has changed as it is now common for part of the inbound validation for on out-of-proc or web request to be invoked.

**Comment by Mordahlhuilhulh** [comment]
Hello @brentschmaltz ,
Just wants to be sure that there isn't any new update regarding the async versions of SignatureProvider.Sign and SignatureProvider.Verify?

**Comment by brentschmaltz** [comment]
@Mordahlhuilhulh not yet ...

**Comment by Mordahlhuilhulh** [comment]
@brentschmaltz  Thanks for your quick replay.
Do you know when the async version is planned?

**Comment by cyberhck** [comment]
@Mordahlhuilhulh it looks like it's planned for V6 looking at assigned milestone.

**Comment by mchandschuh** [comment]
This might help anyone looking to do async JWT validation.

I ended up adding a new interface that also implements `ISecurityTokenValidator`. This allows me to add my custom validator to the `JwtBearerOptions.TokenValidators` collection and in my version of `JwtBearerHandler` (see [#29917](https://github.com/dotnet/aspnetcore/issues/29917)), I check to see if I'm working w/ the synchronous or asynchronous abstraction and act accordingly. The nice part about this approach is that it's a non-breaking change. The not so nice part is the potential for type checks, which I suppose could be hidden away behind an extension method, but still somewhat tacky.

``` csharp
public interface IAsyncSecurityTokenValidator : ISecurityTokenValidator
{
    /// <summary>
    /// Validates a token passed as a string using <see cref="TokenValidationParameters"/>
    /// </summary>
    ValueTask<(ClaimsPrincipal principal, SecurityToken validatedToken)> ValidateTokenAsync(
        string securityToken,
        TokenValidationParameters validationParameters
        );
}

public class CustomAsyncSecurityTokenValidator : JwtSecurityTokenHandler, IAsyncSecurityTokenValidator
{
    // optionally override the synchronous method
    // NOTE: You shouldn't let this actually be called, but it's better that this is called and the real validation happens
    //       rather than running the default `JwtSecurityTokenHandler.ValidateToken` code. You'll need to add custom
    //       IAuthenticationHandler to call the async method appropriately - see `JwtBearerAsyncValidationHandler` in #29917
    public override ClaimsPrincipal ValidateToken(
        string token,
        TokenValidationParameters validationParameters,
        out SecurityToken validatedToken
        )
    {
        ClaimsPrincipal principal;
        (principal, validatedToken) = ValidateTokenAsync(token, validationParameters).AsTask().GetAwaiter().GetResult();
        return principal;
    }

    public ValueTask<(ClaimsPrincipal principal, SecurityToken validatedToken)> ValidateTokenAsync(
        string securityToken,
        TokenValidationParameters validationParameters
        )
    {
        // do async stuff here
    }
}

```

**Comment by stebet** [comment]
> This might help anyone looking to do async JWT validation...

This is a workaround that does semi-work, but the sync-over-async implementation is risky as it might cause threadpool starvation on a high-traffic implementation.

We really need a proper async implementation here, but thanks for this workaround :)

**Comment by mchandschuh** [comment]
> > This might help anyone looking to do async JWT validation...
> 
> This is a workaround that does semi-work, but the sync-over-async implementation is risky as it might cause threadpool starvation on a high-traffic implementation.
> 
> We really need a proper async implementation here, but thanks for this workaround :)

If you look at the linked issue it also includes a custom authentication handler implementation that does a type check to see if the token validator is an `IAsyncSecurityTokenValidator`, so it's async all the way down. There shouldn't be any issues w/ this unless I'm misunderstanding something.

EDIT: Turns out I put the wrong code in the linked issue. I've updated it to properly invoke the `IAsyncSecurityTokenHandler`:
``` csharp
if (validator is IAsyncSecurityTokenValidator asyncValidator)
{
    (principal, validatedToken) = await asyncValidator.ValidateTokenAsync(token, validationParameters).ConfigureAwait(false);
    if (principal == null || validatedToken == null)
    {
        throw new SecurityTokenValidationException("The provided token failed validation.");
    }
}
else
{
    principal = validator.ValidateToken(token, validationParameters, out validatedToken);
}

```

**Comment by HarelM** [comment]
I'm having an issue when I start the server and it stops responding to requests.
The token validation I'm doing is against a third party server (openstreetmap in my case).
I ended up here as I'm concerned what will happen if a lot of users will try to use the server at once.
What I'm doing can be seen here:
https://github.com/IsraelHikingMap/Site/blob/751a30b9c88e3e6dfb52d11605663c45dcc1deac/IsraelHiking.Web/OsmAccessTokenValidator.cs#L56
Which now feels like a great way to get the server stuck... but I'm not sure I know how to properly solve this due to this limitation...

**Comment by goldsam** [comment]
I don't see it mentioned above, but you can currently provide async token validation via the `Events` property of `JwtBearerOptions`. This might provide a stop-gap solution for some users until a proper async `ISecurityTokenValidator` API is implemented:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddAuthentication()
        .AddJwtBearer(jwtBearerOptions =>
        {
            jwtBearerOptions.Events.OnMessageReceived = async (context) =>
            {
                // do some async validation stuff...

                // on happy path:
                context.Principal = ...
                context.Properties = ...;
                context.Success();

                // or sad path:
                context.Fail(...);
            };
        });
}
```

**Comment by HarelM** [comment]
@goldsam Thanks for the info! This seems to be an interesting solution.
This means I don't need to derive from `ISecurityTokenValidator` too, right? Simply supply the relevant code in this section...?
Is there an easy way to avoid sending the request to the third party validator when the same user sends two requests at almost the same time besides locking? which is not advised anymore as far as I understand...? (not talking about caching the response in order to avoid future responses, I mean almost at the same time when the cache does not contain this information - you can see the ugly code I wrote above in the link which just feels wrong now...)

**Comment by HarelM** [comment]
Alongside the `OnMessageRecieved` stuff does the following article and `LockProvider` referenced there would solve the deadlock possibility here:
https://www.ryadel.com/en/asp-net-core-lock-threads-async-custom-ids-lockprovider/
https://github.com/Darkseal/LockProvider/blob/master/LockProvider.cs
For me it seems like the right direction, but I'm not sure If I'm not missing anything related to deadlocks, threads, locks etc...

**Comment by goldsam** [comment]
@HarelM I think locking is a bit out of the scope of this issue, but perhaps you might consider a solution that uses [`Lazy`](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1?view=net-5.0) together with [`ConcurrentDictionary`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-5.0) or a cache of some sort. [LazyCache](https://github.com/alastairtree/LazyCache) might be exactly the solution you are looking for that uses this approach.

**Comment by HarelM** [comment]
@goldsam Thanks for the info! I've implemented the fix according to the article above, it solved the hang.
This lazy cache you referenced is interesting, it might allow me to remove some of the code I added and simplify it.

**Comment by rmandvikar** [comment]
Any eta on this? Without this, same workaround where I'm having to do a blocking call when the cache ttls out or is empty.

**Comment by HarelM** [comment]
@goldsam Thanks for this great suggestion! I have incorporated the lazy cache package in my code and it looks a lot cleaner right now. Thanks!

**Comment by bugproof** [comment]
One of the reasons I stopped using JWT tokens altogether 😄 5+ years for a small change. We might see this change in the next 5 years.

**Comment by brentschmaltz** [comment]
@bugproof it's not a small change and we are close to starting this as it fit well with this PR, which reaches out for metadata in the call graph.

Consider changes related to this PR
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1779/files

**Comment by ricavir11** [comment]
Hi @brentschmaltz, do you have an ETA for this ? 
Our Azure App Service application is struggling because of this :/

**Comment by bugproof** [comment]
@ricavir11 probably in 5 years from now.
Don't see a reason why they couldn't just breaking change it. That wouldn't hurt anyone. Just add an announcement that it's a breaking change like they usually do with many other things.

I completely stopped using anything "Identity" that's coming with ASP.NET Core (along with IdentityUser, IdentityRole and default auth middlewares). I've learned it's a bloated and hard to configure and customize mess. Rolled my own user + auth. Completely dropped usage of JWT tokens as well. Use it only if you care about 3rd party login openid/oauth stuff. Auth solution that comes with ASP.NET Core is just really bad IMHO. JWTs shouldn't be used for user sessions as well - which is a mistake I made early on due to how much hype there was around it (you can't have stateless auth with a stateful backend)

**Comment by ricavir11** [comment]
@bugproof, I hope this will not be so long 🤦‍♂️ !
I'm using it along with aspnetzero framework and it is working well for low traffic apps.
As soon as you get more hits to the server, things become bad ! I'm having huge waiting timeouts just to validate tokens .... like 20s or even more. Adding more CPU or RAM does not change the problem.
I really don't understand why this issue has not been put in the top priority of Microsoft team.

**Comment by brentschmaltz** [comment]
@bugproof there are a couple of items ahead of this. I believe we will get this in 2022

**Comment by benmccallum** [comment]
Got an email from Auth0 today that some of our signing keys are old and it's best practice that they are rotated out. (You rotate in a "next" to become "current" and then will effectively have two keys that are active until you can fully revoke the old one once you're happy all the JWTs signed with it are expired).

So we need a way our apps can fall through at runtime failure (with current known keys) and go grab the latest from Auth0 asyncly, then try "one more time". [Source](https://auth0.com/docs/secure/tokens/json-web-tokens/locate-json-web-key-sets). 

It'd be great if this was a scenario that was easily supported + documented. i.e. it'll be great to have the token grabbing support async, that'll unblock some scenarios; but a further feature to setup a recall of `IssuerSigningKeyResolver` and a retry of validation once would be incredible!

Right now I think the only choice we have is to rotate the old one, then immediately force our apps to grab the latest keys again somehow. Is there anything that can be done in an event (that's async) now, that would support grabbing the new keys in a way that'd loop back for another attempt at validation?

**Comment by brentschmaltz** [comment]
@benmccallum @ricavir11 we added the first part of our Async effort in this PR: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1810

@benmccallum This PR may interest you: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1797 

Enabling the feature requires [TokenValidationParameters.BaseConfigurationManager](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/adc8d59d4aaae502f77fd8f758e4f020045f69eb/src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs#L416) to be set.

When enabled the following logic will apply:

If JWT has a 'kid' and if the JWT is not {expired, some other cases, ...}, reach out and get new metadata and try again.
If the new metadata still results in a failure, the last metadata that resulted in a successful validation was 'remembered' as LastKnownGood and will be re-used for a default of 1 hour. This protects against the publication of bad metadata.

The feature can be disabled and will only work if TokenValidationParameters.BaseConfigurationManager is set, so existing users will not see any effects. 

We are working with the asp.net team to get this into asp.net 7.0

**Comment by rmandvikar** [comment]
@brentschmaltz That's a good start, and thanks for the update. What's the plan/eta to make `JwtSecurityTokenHandler.ValidateTokenAsync()` method fully async? Currently, it's just a wrapper around the sync `ValidateToken()` method. 

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/adc8d59d4aaae502f77fd8f758e4f020045f69eb/src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs#L1654-L1675

**Comment by brentschmaltz** [comment]
@rmandvikar i know it has been on the list for a while and we finally inched towards this gap.

We have a couple of items before we can think about getting this into a release.
This is mainly about having Asp.Net 7.0 using 6.x as the default and use JsonWebTokenHandler / JsonWebToken 

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1726
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1654
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1426
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1197
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1160

Using System.Text.Json, that does away with all our splitting and translating from string -> bytes -> string
as in: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1805

**Comment by mendhak** [comment]
> Right now I think the only choice we have is to rotate the old one, then immediately force our apps to grab the latest keys again somehow. Is there anything that can be done in an event (that's async) now, that would support grabbing the new keys in a way that'd loop back for another attempt at validation?

@benmccallum  would that mean simply redeploying the applications, so the caches are cleared?

**Comment by benmccallum** [comment]
@mendhak , yea, or via some admin screen that triggers it, but it's not exactly a clean solution.

**Comment by brentschmaltz** [comment]
@benmccallum @mendhak are you using ConfigurationManager to obtain metadata?
ConfigurationManager supports obtaining metadata on a schedule and one-off refresh.

Obtaining metadata is async.

**Comment by benmccallum** [comment]
@brentschmaltz , I don't believe we are. It's been a while since I was looking at this and haven't had a chance to revisit.

**Comment by wilsoncg** [comment]
Hello, what's happened to this issue?
It seems there's been no activity for a [year](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/468#issuecomment-1067272989).

**Comment by jennyf19** [comment]
This is done as part of IdentityModel 7x, please try with latest version and reopen if necessary.

**Comment by HarelM** [comment]
Can you link to a demo/documentation of this new ability?

**Comment by bugproof** [comment]
Pro tip: please don't close issues without linking to the commit/demo/documentation

This is still an issue without a proper message of what is done

**Comment by nathfy** [comment]
I think this is missing info: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/IdentityModel-7x#sync-over-async

"ValidateToken was a sync method calling an async method, which can lead to [threadpool starvation when on the hot path](https://devblogs.microsoft.com/pfxteam/should-i-expose-synchronous-wrappers-for-asynchronous-methods/), this method has been deprecated, please use ValidateTokenAsync instead. See issue [#2253](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2253) for details."

Also note this issue when upgrading the nugets: https://stackoverflow.com/a/77121911

**Comment by ownerer** [comment]
I'm sorry but I don't see how this issue is actually solved?
Looking at the [source code for JwtSecurityTokenHandler](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs) all this new `ValidateTokenAsync` method does is wrap the old synchronous method.
Everything else is still synchronous including delegates in the `TokenValidationParameters` class, such as `IssuerSigningKeyResolver`.
I'd expect to be able to pass an asynchronous delegate there; that is still impossible.
So getting the signing key from some external resource still has to happen sync-over-async...
Eg:

```
IssuerSigningKeyResolver = (token, securityToken, kid, parameters) =>
{
    var result = new List<SecurityKey>();
    //TODO: make this async, cfr https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/468
    var secret = _tokenManager.GetSecret(kid, cancellationToken).Result; // <<<---- BAD!!
    if (secret != null)
    {
        result.Add(new SymmetricSecurityKey(Encoding.ASCII.GetBytes(secret.Value)));
    }

    return result;
}
```

**Comment by kroymann** [comment]
@ownerer All of the async improvements appear to have gone into `JsonWebTokenHandler`, which is apparently the new and improved successor to `JwtSecurityTokenHandler` (see https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/945#issuecomment-405113306 - I sure wish this was better documented as I had no idea `JwtSecurityTokenHandler` was old until just now).  It looks they did the minimal update necessary to have a technically functional implementation of the async methods for `JwtSecurityTokenHandler` (https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1810/files#diff-0b1054c790c38d665f14eee85e63889a6c64f63e87ab7ded11b86708f1ed1475), but if you want a proper async-all-the-way-down option, you need to switch to `JsonWebTokenHandler`.

**Comment by ownerer** [comment]
Thanks for that info @kroymann , but the `ValidateTokenAsync` method of `JsonWebTokenHandler` still takes an instance of that same `TokenValidationParameters` type, meaning its delegates are still synchronous, so the problem I described remains. You can see how it gets called [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/3cae45521ede8fb4c77f9be73887b4818f756f9f/src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs#L1849).
This issue is not resolved.

**Comment by brentschmaltz** [comment]
@ownerer we are aware of the issue with the sync delegates.
We tried out one async, https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/3cae45521ede8fb4c77f9be73887b4818f756f9f/src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs#L115 to get a feel for the design.

**Comment by ownerer** [comment]
@brentschmaltz , aha ok.
Any idea when the others will be converted/added?
Is that work tracked somewhere, seeing as this issue is closed...

**Comment by brentschmaltz** [comment]
@ownerer we are fully booked until the new year, but new async delegates that return a typeof ValidationResult to avoid logging or throwing when multiple identity providers are in play and the second one succeeds, is important work that we are planning early next year.

**Comment by Pikhulya** [comment]
_>... This is done as part of IdentityModel 7x, please try with latest version and reopen if necessary. ..._

@jennyf19 , @brentschmaltz , please re-open the issue given, as outlined in the few above comments, it is not really fixed. Alternatively, if it is tracked by another tracking work item on GitHub - please share the link.

**Comment by brentschmaltz** [comment]
@Pikhulya reopened

**Comment by HarelM** [comment]
Are there any examples I can use in order to change my code to use `JsonWebTokenHandler`? I can't seem to find any information and I'm not able to wire it up to my app so that it does the validation...

**Comment by brentschmaltz** [comment]
@HarelM marked this a .net9 as we are going to be modifying our Delegates to make them async and return a ValidationResult rather than throwing.

**Comment by HarelM** [comment]
Yeah, I'm not going to wait for .net 9 just to find out the documentation is poor again and that I can't figure out how to properly migrate my current code.
While in most cases I disagree with the approach of "if it ain't broken don't fix it", in this case I gave up.
Here's my commit BTW that only fixes what I needed in terms of validating the token only for authorized routes
https://github.com/IsraelHikingMap/Site/commit/bae22472aab6af88716ae740a494ea97e334d2b1

**Comment by jennyf19** [comment]
@HarelM we will aim to get this into our current workstream. by referring to .NET 9, we do not mean Nov 2024. Please follow this issue for updates. Appreciate the feedback and we will do better with documentation moving forward.

**Comment by jennyf19** [comment]
related issue: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2711

**Comment by bugproof** [comment]
So are we going to get this until 2077? 😂 😂 😂 😂 😂 😂 😂 

Maybe in .NET 20. Sorry won't make it for .NET 10

**Comment by PascalVorwerk** [comment]
Hi, is there any news on this issue? We are trying to validate a JWT we are sending in a B2B integration, without having a full OIDC provider. So we need to manually retrieve the public keys. We would like to do this through a JWKS endpoint, but this requires network traffic and so it requires async handling. I googled around and figured there are some workarounds, but i am wondering if there is something official for it?
